---
title: "heatmap and volcano"
author: "Howard Cen"
date: "2024-03-23"
output: html_document
---

```{r}
#if (!requireNamespace("BiocManager", quietly = TRUE))
#  install.packages("BiocManager")
#BiocManager::install("DESeq2")
#library("DESeq2")

#BiocManager::install("org.Hs.eg.db")
#library(org.Hs.eg.db)

library(tidyverse)
library(ggrepel)

library(readxl) # for reading excel file
library(ComplexHeatmap)
library(circlize) # installed with ComplexHeatmap, but need to load it
library(gridtext) # needed for complex heatmap
library(scales) # for calculating z-scores etc

setwd(dirname(rstudioapi::getActiveDocumentContext()$path)) ##Set working directory to where this file is.
getwd()



```


```{r function plot.volcano}
library(tidyverse)
library(ggrepel)
library(plotly)

plot.volcano <- function(df, p_col, p_cutoff, log2FC_col, log2FC_cutoff, gene_num, gene_col, name){
  
  df <- df %>% 
    mutate(log_padj = -log10(!!rlang::sym(p_col)))
#           sign_p = -log10(!!rlang::sym(p_col)) * sign(!!rlang::sym(log2FC_col)))
  
  # select top or neg x genes to label
  pos_label <- df[order(df[[p_col]]), ] %>% 
    filter(!!rlang::sym(log2FC_col) > 0) %>%
    slice_head(n = gene_num) %>%
    pull(gene_col)
  
  neg_label <- df[order(df[[p_col]]), ] %>% 
    filter(!!rlang::sym(log2FC_col) < 0) %>%
    slice_head(n = gene_num) %>%
    pull(gene_col)

  label_genes <- c(pos_label,neg_label)
  
  df <- df %>%
    mutate(label = ifelse(!!rlang::sym(gene_col) %in% label_genes, !!rlang::sym(gene_col), ""))
  
  write.csv(df, paste0("output/volcano_", name, ".csv"), row.names = F)
  
  df <- df %>%
    mutate(label = ifelse(label %in% label_genes, paste0("italic('", label, "')"), NA))

  # manually made Inf values 1.1x max value !!!
  df.inf <- df
  df.inf[!is.finite(df$log_padj), "log_padj"] <- max(df$log_padj[is.finite(df$log_padj)]) * 1.1
  
  df_pos <- df.inf[df.inf[[log2FC_col]] > log2FC_cutoff & df.inf[[p_col]] < p_cutoff,]
  df_neg <- df.inf[df.inf[[log2FC_col]] < -log2FC_cutoff & df.inf[[p_col]] < p_cutoff,]

  df_ns <- df.inf[!df.inf[[gene_col]] %in% c(df_pos[[gene_col]], df_neg[[gene_col]]), ]


    
  p <- ggplot(data=df.inf, #[plot.df$adj.p<0.05,],
           aes(x = !!rlang::sym(log2FC_col), y = log_padj 
               )) +
    ylab(expression('-log10 adj.Pvalue')) + #-log[10]~(adj.~P~value)
    xlab(expression("log2 fold change")) +
    geom_point(data = df_ns,
               size = 2.5,
               color = "grey", stroke = 0, alpha = 0.6) +
    geom_point(data = df_pos,
               size = 2.5,
               color = "red4", stroke = 0, alpha = 0.6) +
    geom_point(data = df_neg,
               size = 2.5,
               color = "blue4", stroke = 0, alpha = 0.6) +
    #geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
    
    #ylim(c(NA, max((-log10( df[is.finite(-log10(df[[p_col]])),][[p_col]])))*1.3)) +

    ggrepel::geom_text_repel(
      parse = TRUE, # make it recognize the italic fonts of gene names
        data = df_pos, 
        aes(label = label),
        color="red4",
        size=4,
        nudge_x =  max(df[[log2FC_col]]) + max(abs(min(df[[log2FC_col]])),abs(max(df[[log2FC_col]])))*0.5 - df_pos[[log2FC_col]],
        segment.size=0.1, 
        #segment.color="grey", 
        direction="y", 
        hjust= 1, # adjusting this somehow reduced overlap
        #max.iter = 100000,
        max.overlaps = Inf) +
    
   ggrepel::geom_text_repel(
     parse = TRUE, # make it recognize the italic fonts of gene names
        data = df_neg, 
        aes(label = label),
        color="blue4",
        size=4,
        nudge_x =  min(df[[log2FC_col]]) - max(abs(min(df[[log2FC_col]])),abs(max(df[[log2FC_col]])))*0.5 - df_neg[[log2FC_col]],
        segment.size=0.1, 
        #segment.color="grey", 
        direction="y", 
        hjust= 0, # adjusting this somehow reduced overlap
        #max.iter = 100000,
        max.overlaps = Inf) +
    
    #ggtitle(name) +
    
    coord_cartesian(clip = "off") +
    
      theme(legend.position = "right",
            legend.title = element_text(hjust = 0, # left align
                                        margin = margin(t = 15, unit = "pt")) # add more space on top of legend titles
            #legend.spacing.y = unit(1,"cm")
            ) +
      theme(panel.grid.minor.y = element_blank(),
            panel.grid.minor.x = element_blank(),
            text=element_text(size=14),
            axis.text=element_text(size=12),
            axis.title=element_text(size=14),
            legend.text=element_text(size=14),
            legend.title=element_text(size=14),
            aspect.ratio = 1/1.2, panel.grid.major = element_blank(),
            panel.background = element_blank(), axis.line = element_line(colour = "black")) + 
      theme(plot.title = element_text(hjust = 0.5, face = "italic", colour="grey30", size=15))
  
  fpath <- paste0("figures/volcano_", name)
  
  p
  ggsave(paste0(fpath,".pdf"), height = 5.5, width = 6.5, units = "in")
  
  p
  ggsave(paste0(fpath,".svg"), device = svg, height = 5.5, width = 6.5, units = "in")
  
  p1 <- ggplot(data=df.inf, #[plot.df$adj.p<0.05,],
           aes(x=!!rlang::sym(log2FC_col), y=log_padj,
               text = paste0("Gene: ", symbol)
               )) +
    ylab('-log10 adj.Pvalue') + #-log[10]~(adj.~P~value)
    xlab("log2 fold change") +
    geom_point(data = df_ns,
               size = 2.5,
               color = "grey", stroke = 0, alpha = 0.6) +
    geom_point(data = df_pos,
               size = 2.5,
               color = "red4", stroke = 0, alpha = 0.6) +
    geom_point(data = df_neg,
               size = 2.5,
               color = "blue4", stroke = 0, alpha = 0.6) +
    #geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
    
    #ylim(c(NA, max((-log10( df[is.finite(-log10(df[[p_col]])),][[p_col]])))*1.3)) +
    
    #ggtitle(name) +
    
    coord_cartesian(clip = "off") +
    
      theme(legend.position = "right",
            legend.title = element_text(hjust = 0, # left align
                                        margin = margin(t = 15, unit = "pt")) # add more space on top of legend titles
            #legend.spacing.y = unit(1,"cm")
            ) +
      theme(panel.grid.minor.y = element_blank(),
            panel.grid.minor.x = element_blank(),
            text=element_text(size=14),
            axis.text=element_text(size=12),
            axis.title=element_text(size=14),
            legend.text=element_text(size=14),
            legend.title=element_text(size=14),
            panel.grid.major = element_blank(),
            panel.background = element_blank(), axis.line = element_line(colour = "black")) + 
      theme(plot.title = element_text(hjust = 0.5, face = "italic", colour="grey30", size=15))

  
    #htmlwidgets::saveWidget(partial_bundle(ggplotly(p1)), "figures/p11.html")  #https://plotly-r.com/saving
    htmlwidgets::saveWidget(ggplotly(p1), paste0(fpath,".html"), selfcontained = F, libdir = "plotly_lib")
    
    return(p)
}

```

```{r plot volcano}
de0 <- read.csv("output/RNAseq_DESeq2_eIF2A0_control0_noOutlier_lfc0.5.csv")
View(de0)
plot.volcano(df = de0,
             p_col = 'padj',
             p_cutoff = 0.05,
             log2FC_col = 'log2FoldChange',
             log2FC_cutoff = 1,
             gene_num = 30,
             gene_col = 'symbol',
             name = 'RNAseq_Eif2a0_ctrl0')

```

```{r heatmap RNAseq processing}
#e0_c0 <- read.csv("output/RNAseq_DESeq2_eIF2A0_control0_noOutlier_lfc0.5.csv")

c1_c0 <- read.csv("output/RNAseq_DESeq2_control1_control0_noOutlier_lfc0.5.csv")

e1_c1 <- read.csv("output/RNAseq_DESeq2_eIF2A1_control1_noOutlier_lfc0.5.csv")

e1_e0 <- read.csv("output/RNAseq_DESeq2_eIF2A1_eIF2A0_noOutlier_lfc0.5.csv")

#

c3_c0 <- read.csv("output/RNAseq_DESeq2_control3_control0_noOutlier_lfc0.5.csv")

e3_c3 <- read.csv("output/RNAseq_DESeq2_eIF2A3_control3_noOutlier_lfc0.5.csv")

e3_e0 <- read.csv("output/RNAseq_DESeq2_eIF2A3_eIF2A0_noOutlier_lfc0.5.csv")

#

c1_c0_sig <- filter(c1_c0, padj<0.05)

e1_c1_sig <- filter(e1_c1, padj<0.05)

e1_e0_sig <- filter(e1_e0, padj<0.05)

View(c1_c0_sig)
View(e1_c1_sig)

#

c3_c0_sig <- filter(c3_c0, padj<0.05)

e3_c3_sig <- filter(e3_c3, padj<0.05)

e3_e0_sig <- filter(e3_e0, padj<0.05)

View(c3_c0_sig)
View(e3_c3_sig)

#

gene_c0_c1_e1 <- intersect(c1_c0_sig$symbol, e1_c1_sig$symbol)

c0_c1_e0_e1 <- intersect(c1_c0_sig$symbol, e1_e0_sig$symbol)

#

gene_c0_c3_e3 <- intersect(c3_c0_sig$symbol, e3_c3_sig$symbol)

c0_c3_e0_e3 <- intersect(c1_c0_sig$symbol, e1_e0_sig$symbol)

#

de_c0_c1_e1 <- c1_c0 %>% inner_join(e1_c1, by = c("symbol","entrez","fullname"), suffix = c(".c1c0", ".e1c1")) %>%
  filter(symbol %in% gene_c0_c1_e1) %>%
  mutate(direction = case_when(log2FoldChange.c1c0>0 & log2FoldChange.e1c1>0 ~ "Up",
                               log2FoldChange.c1c0<0 & log2FoldChange.e1c1<0 ~ "Down",
                               .default = "Reverse"))
dim(de_c0_c1_e1) # 34

de_c0_c3_e3 <- c3_c0 %>% inner_join(e3_c3, by = c("symbol","entrez","fullname"), suffix = c(".c3c0", ".e3c3")) %>%
  filter(symbol %in% gene_c0_c3_e3) %>%
  mutate(direction = case_when(log2FoldChange.c3c0>0 & log2FoldChange.e3c3>0 ~ "Up",
                               log2FoldChange.c3c0<0 & log2FoldChange.e3c3<0 ~ "Down",
                               .default = "Reverse"))
dim(de_c0_c3_e3) # 242

View(de_c0_c1_e1)
View(de_c0_c3_e3)
genes <- unique(c(de_c0_c1_e1[de_c0_c1_e1$direction == "Reverse",]$symbol, de_c0_c3_e3[de_c0_c3_e3$direction == "Reverse",]$symbol))

de_c0c1e1_c0c3e3 <- c1_c0 %>% 
  inner_join(e1_c1, by = c("symbol","entrez","fullname"), suffix = c("", ".e1c1")) %>% 
  inner_join(c3_c0, by = c("symbol","entrez","fullname"), suffix = c("",".c3c0")) %>%
  inner_join(e3_c3, by = c("symbol","entrez","fullname"), suffix = c(".c1c0", ".e3c3")) %>% 
  filter(symbol %in% genes) 

View(de_c0c1e1_c0c3e3)

de_c0c1e1_c0c3e3_ord <- de_c0c1e1_c0c3e3[order(rowSums(de_c0c1e1_c0c3e3[,grep("padj", colnames(de_c0c1e1_c0c3e3))])), ]

de_c0c1e1_c0c3e3_ord.e <- de_c0c1e1_c0c3e3[order(rowSums(de_c0c1e1_c0c3e3[,grep("padj.e", colnames(de_c0c1e1_c0c3e3))])), ]

de_c0c1e1_c0c3e3_ord.tg <- de_c0c1e1_c0c3e3[order(de_c0c1e1_c0c3e3$padj.c3c0),]

de_c0c1e1_c0c3e3_ord.tg.e <- de_c0c1e1_c0c3e3[order(de_c0c1e1_c0c3e3$padj.c3c0 + de_c0c1e1_c0c3e3$padj.e3c3),]

dim(de_c0c1e1_c0c3e3_ord) #142
dim(de_c0c1e1_c0c3e3_ord.e)
dim(de_c0c1e1_c0c3e3_ord.tg)

write.csv(de_c0c1e1_c0c3e3_ord, "output/heatmap_EIF2A_RNAseq_reverse_order.csv", row.names = F)

write.csv(de_c0c1e1_c0c3e3_ord.e, "output/heatmap_EIF2A_RNAseq_reverse_order_eif2a.csv", row.names = F)

write.csv(de_c0c1e1_c0c3e3_ord.tg, "output/heatmap_EIF2A_RNAseq_reverse_order_Tg3h.csv", row.names = F)

write.csv(de_c0c1e1_c0c3e3_ord.tg.e, "output/heatmap_EIF2A_RNAseq_reverse_order_Tg3h_eif2a.csv", row.names = F)

```


```{r heatmap all}
norm.data <- read.csv("output/norm_vst.csv", row.names = 1)
View(norm.data)

cname <- colnames(norm.data)
cname_ord <- cname[c(grep("control0", cname),grep("eIF2A0", cname),grep("control1", cname), grep("eIF2A1", cname), grep("control3", cname),grep("eIF2A3", cname))]

norm.data <- norm.data[, cname_ord]

genes <- unique(c(de_c0_c1_e1[de_c0_c1_e1$direction == "Reverse",]$symbol, de_c0_c3_e3[de_c0_c3_e3$direction == "Reverse",]$symbol))

m <- norm.data[genes,]

m <- norm.data[de_c0c1e1_c0c3e3_ord$symbol[1:50], ]

m <- norm.data[de_c0c1e1_c0c3e3_ord.e$symbol[1:50], ]

m <- norm.data[de_c0c1e1_c0c3e3_ord.tg$symbol[1:50], ]

m <- norm.data[de_c0c1e1_c0c3e3_ord.tg.e$symbol[1:50], ]

m <- norm.data[de_c0c1e1_c0c3e3_ord.tg.e$symbol[1:50], c(1:3, 13:15, 4:6, 16:18)]

m.z <- t(scale(t(m))) # scale -- calculate Z-score for columns: [(x-mean)/SD]; t -- transpose the matrix; log2 -- log2 transform the data
colnames(m.z)
max(m.z)
min(m.z)
View(m.z)

#rownames(m.z) <- rownames(m) # set gene names as rownames of m.z

# plot heatmap  --------------
# (originally established by Yi Han Xia)

number_c0 <- length(grep("control0", colnames(m.z))) # grep() -- find the position/index of, in this case, column names that contain the pattern "DN"
number_c1 <- length(grep("control1", colnames(m.z)))
number_c3 <- length(grep("control3", colnames(m.z)))

number_e0 <- length(grep("eIF2A0", colnames(m.z))) # grep() -- find the position/index of, in this case, column names that contain the pattern "DN"
number_e1 <- length(grep("eIF2A1", colnames(m.z)))
number_e3 <- length(grep("eIF2A3", colnames(m.z)))

end_c0 <- grep("control0", colnames(m.z))[number_c0] # get the end position/index of each group
end_c1 <- grep("control1", colnames(m.z))[number_c1]
end_c3 <- grep("control3", colnames(m.z))[number_c3]

end_e0 <- grep("eIF2A0", colnames(m.z))[number_e0]
end_e1 <- grep("eIF2A1", colnames(m.z))[number_e1]
end_e3 <- grep("eIF2A3", colnames(m.z))[number_e3]


#set.seed(12) # run this to get the same k-means clusters 
cbPalette <- c("#CC79A7", #magenta"
               "#E69F00", #lightorange
               "#D55E00", #darkorange
               
               "#009E73", #green
               "#56B4E9", #blue
               "#0072B2", #darkblue
               "#F0E442", #yellow
               "#999999" #grey
)

# heatmap for DMSO_EIF2A vs DMSO or Tg_EIF2A vs Tg                
heatmap <- Heatmap(m.z,
                   show_row_names = TRUE,#FALSE,
                   show_column_names = FALSE,
                   show_row_dend = TRUE,
                   show_column_dend = FALSE,
                   
                   clustering_distance_rows = "euclidean",
                   clustering_method_rows = "ward.D2", # default linkage method is "complete"
                   
                   row_dend_side = "left",
                   row_dend_width = unit(15, "mm"),
                   
                   column_names_side = "top",
                   column_dend_side = "bottom",
                   
                   row_labels = gt_render(rownames(m.z)), ### now we set row names (genes) as labels
                   row_names_gp = gpar(fontsize = 10),
                  
                   col = colorRamp2(c(min(m.z),0,max(m.z)), c("blue", "white", "red")), # set the color range from Z-score -3 to 3 (saturate the outliers,  if you didn't remove them).
                   #col = colorRamp2(c(min(na.omit(m.z)),0,max(na.omit(m.z))), c("blue", "white", "red")), # alternatively, set the range from min to max value
                   
                   column_order = 1:ncol(m.z),
                   
                   height = unit(220, "mm"), # set the height of heatmap
                   width = ncol(m.z)*unit(8, "mm"), # set the column width
                   
                   top_annotation = columnAnnotation(empty = anno_empty(border = FALSE, 
                                                                        height = unit(6, "mm"))), # leave an empty space on top for the condition labels below.
                   
#                   right_annotation = 
#                           rowAnnotation(`DMSO_EIF2A vs DMSO` = de$sig_DMSO_EIF2A.ctrl,
#                                         `Tg_EIF2A vs Tg` = de$sig_Tg_EIF2A.ctrl,
                                         
                                         
#                                         col = list(`DMSO_EIF2A vs DMSO` = c("TRUE" = cbPalette[1]),
#                                                    `Tg_EIF2A vs Tg` = c("TRUE" = cbPalette[2])
#                                         ),
#                                         annotation_legend_param = list(
                      
#                                           `DMSO_EIF2A vs DMSO` = list(
#                                             title = c(""),
#                                             labels = "DMSO_EIF2A vs DMSO"#expression(italic("Ins1")^italic("+/+")*" Sucrose vs Water") #expression() issue - https://github.com/jokergoo/ComplexHeatmap/issues/678
#                                           ),
                      
#                                           `Tg_EIF2A vs Tg` = list(
#                                             title = c(""),
#                                             labels = "Tg_EIF2A vs Tg"
#                                           )
#                                         ),
#                                         show_annotation_name = FALSE,
#                                         show_legend=TRUE,
#                                         na_col = "white",
#                                         width = unit(8, "mm")
#                      
#                           ),

                   
                   
                   border_gp = gpar(col = "black"), # set border color
                   
                   show_heatmap_legend = TRUE,
                   heatmap_legend_param = list(
                     title = "Row z-score",
                     title_position = "topleft",
                     #at = c(-3, -2, -1, 0, 1, 2, 3), # set ticks/numbers of legend
                     legend_height = unit(4, "cm")),
                   
                   #split = 4, # split the heatmap into separate blocks by hierarchical clustering
                   #row_km = 4, # or split the heatmap into separate blocks by k-means clustering
                   ## determine the number of clusters: https://www.datanovia.com/en/lessons/determining-the-optimal-number-of-clusters-3-must-know-methods/
                   
                  layer_fun = function(j, i, x, y, width, height, fill) {
                     mat = restore_matrix(j, i, x, y)
                     
                     ind = unique(c(mat[, c(end_c0,
                                            end_e1,
                                            end_c1,
                                            end_e1,
                                            end_c3
                                            ) # enter the number where you want a gap between columns
                                        ])) 
                     
                     grid.rect(x = x[ind] + unit(0.5/ncol(m.z), "npc"), 
                               y = y[ind], 
                               width = unit(0.03, "inches"), # width of the gap
                               height = unit(1/nrow(m.z), "npc"),
                               gp = gpar(col = "white", fill = "white") # color of the gap
                     )
                   }
                   ) 

draw(heatmap)

# save and label heatmap ---------------

#step 1 - save a pdf file
{pdf(file = 'figures/heatmap_EIF2A_RNAseq_reverse_all.pdf',
    width = 8.5, 
    height = 10)
  
{pdf(file = 'figures/heatmap_EIF2A_RNAseq_reverse_top50.pdf',
    width = 8.5, 
    height = 10)

{pdf(file = 'figures/heatmap_EIF2A_RNAseq_reverse_top50eif2a.pdf',
    width = 8.5, 
    height = 10)

{pdf(file = 'figures/heatmap_EIF2A_RNAseq_reverse_top50Tg3h.pdf',
    width = 8.5, 
    height = 10)

{pdf(file = 'figures/heatmap_EIF2A_RNAseq_reverse_top50.Tg3h.eif2a.pdf',
    width = 8.5, 
    height = 10)


#step 2 - draw heatmap in the saved file
draw(heatmap)

#step 3 - draw the labeling on the heatmap in the saved file

# labeling heatmap
#list_components() # get the viewport names
seekViewport("annotation_empty_1") # seek the empty space we saved at the top of heatmap, called "annotation_empty_1"

## draw text labels
grid.text(expression("Control\nTg_0h"), 
          x = (end_c0)/2/ncol(m.z), # location at the middle of DN
          y = 0.5,
          just = c("center", "center"),
          gp = gpar(fontsize = 11,
                    col = "black"))
?grid.text
grid.text(expression("EIF2A\nTg_0h"), 
          x = (end_c0 + end_e0)/2/ncol(m.z),
          y = 0.5,
          just = c("center", "center"),
          gp = gpar(fontsize = 11))

grid.text(expression("Control\nTg_1h"), 
          x = (end_e0 + end_c1)/2/ncol(m.z), # location at the middle of DH
          y = 0.5,
          just = c("center", "center"),
          gp = gpar(fontsize = 11))

grid.text(expression("EIF2A\nTg_1h"), 
          x = (end_c1 + end_e1)/2/ncol(m.z),
          y = 0.5,
          just = c("center", "center"),
          gp = gpar(fontsize = 11))

grid.text(expression("Control\nTg_3h"), 
          x = (end_e1 + end_c3)/2/ncol(m.z),
          y = 0.5,
          just = c("center", "center"),
          gp = gpar(fontsize = 11))

grid.text(expression("EIF2A\nTg_3h"),
          x = (end_c3 + end_e3)/2/ncol(m.z),
          y = 0.5,
          just = c("center", "center"),
          gp = gpar(fontsize = 11))

dev.off()
}


{pdf(file = 'figures/heatmap_EIF2A_RNAseq_reverse_top50.Tg3h.eif2a_rearranged.pdf',
    width = 7, 
    height = 10)
#step 2 - draw heatmap in the saved file
draw(heatmap)

#step 3 - draw the labeling on the heatmap in the saved file

# labeling heatmap
#list_components() # get the viewport names
seekViewport("annotation_empty_1") # seek the empty space we saved at the top of heatmap, called "annotation_empty_1"

## draw text labels
grid.text(expression("Control\nTg_0h"), 
          x = (end_c0)/2/ncol(m.z), # location at the middle of DN
          y = 0.5,
          just = c("center", "center"),
          gp = gpar(fontsize = 11,
                    col = "black"))

grid.text(expression("Control\nTg_3h"), 
          x = (end_c0 + end_c3)/2/ncol(m.z),
          y = 0.5,
          just = c("center", "center"),
          gp = gpar(fontsize = 11))

grid.text(expression("EIF2A\nTg_0h"), 
          x = (end_c3 + end_e0)/2/ncol(m.z),
          y = 0.5,
          just = c("center", "center"),
          gp = gpar(fontsize = 11))

grid.text(expression("EIF2A\nTg_3h"),
          x = (end_e0 + end_e3)/2/ncol(m.z),
          y = 0.5,
          just = c("center", "center"),
          gp = gpar(fontsize = 11))

dev.off()
}
```

```{r heatmap proteomics processing}
#e0_c0 <- read.csv("output/RNAseq_DESeq2_eIF2A0_control0_noOutlier_lfc0.5.csv")
library(readxl)
dmso.e_dmso.c <- read_excel("output/EIF2A_proteomics_limma_DE_results.xlsx", sheet = 1)
tg.e_tg.c <- read_excel("output/EIF2A_proteomics_limma_DE_results.xlsx", sheet = 2)
tg.c_dmso.c <- read_excel("output/EIF2A_proteomics_limma_DE_results.xlsx", sheet = 3)
tg.e_dmso.e <- read_excel("output/EIF2A_proteomics_limma_DE_results.xlsx", sheet = 4)

View(dmso.e_dmso.c)
#
library(tidyverse)
dmso.e_dmso.c_sig <- filter(dmso.e_dmso.c, adj.P.Val<0.05)
tg.e_tg.c_sig <- filter(tg.e_tg.c, adj.P.Val<0.05)
tg.c_dmso.c_sig <- filter(tg.c_dmso.c, adj.P.Val<0.05)
tg.e_dmso.e_sig <- filter(tg.e_dmso.e, adj.P.Val<0.05)

tg.e_tg.c_unadj <- filter(tg.e_tg.c, P.Value<0.001)

View(tg.e_dmso.e_sig)

#

gene_dmso.c_tg.c_tg.e <- intersect(tg.c_dmso.c_sig$symbol, tg.e_tg.c_sig$symbol)


dmso.c_tg.c_tg.e <- tg.c_dmso.c_sig %>% inner_join(tg.e_tg.c_sig, by = c("symbol","entrez","ENSG","fullname"), suffix = c("_Tg.Ctrl_DMSO.Ctrl", "_Tg.DMSO_Tg.Ctrl")) %>%
  filter(symbol %in% gene_dmso.c_tg.c_tg.e) %>%
  mutate(direction = case_when(logFC_Tg.Ctrl_DMSO.Ctrl>0 & logFC_Tg.DMSO_Tg.Ctrl>0 ~ "Up",
                               logFC_Tg.Ctrl_DMSO.Ctrl<0 & logFC_Tg.DMSO_Tg.Ctrl<0 ~ "Down",
                               .default = "Reverse"))
View(dmso.c_tg.c_tg.e)

dmso.c_tg.c_tg.e_unadj <- tg.c_dmso.c_sig %>% inner_join(tg.e_tg.c_unadj, by = c("symbol","entrez","ENSG","fullname"), suffix = c("_Tg.Ctrl_DMSO.Ctrl", "_Tg.DMSO_Tg.Ctrl")) %>%
  filter(symbol %in% gene_dmso.c_tg.c_tg.e) %>%
  mutate(direction = case_when(logFC_Tg.Ctrl_DMSO.Ctrl>0 & logFC_Tg.DMSO_Tg.Ctrl>0 ~ "Up",
                               logFC_Tg.Ctrl_DMSO.Ctrl<0 & logFC_Tg.DMSO_Tg.Ctrl<0 ~ "Down",
                               .default = "Reverse"))
View(dmso.c_tg.c_tg.e_unadj)

write.csv(dmso.c_tg.c_tg.e, "output/EIF2A_proteomics_limma_DE_reverse.csv", row.names = F)


dmso.c_tg.c_tg.e <- read.csv("output/EIF2A_proteomics_limma_DE_reverse.csv")
dmso.c_tg.c_tg.e_0.25 <- read.csv("output/EIF2A_proteomics_limma_DE_reverse_0.25detected.csv")

```

```{r heatmap proteomics plotting}

protein_data_wide <- read.csv("output/EIF2A_proteomics_log2_norm_impute.csv", row.names = 1)
protein_data_wide_non_imputed <- read.csv("output/EIF2A_proteomics_log2_norm_nonimpute.csv", row.names = 1)

m <- protein_data_wide_non_imputed[gene_dmso.c_tg.c_tg.e,]

m <- protein_data_wide_non_imputed[dmso.e_dmso.c_sig$symbol,]
m <- protein_data_wide_non_imputed[tg.e_tg.c_sig$symbol,]
m <- protein_data_wide_non_imputed[tg.c_dmso.c_sig$symbol,]
m <- protein_data_wide_non_imputed[tg.e_dmso.e_sig$symbol,]

# plot significant proteins, without imputed values
heatmap_proteomics(dat = protein_data_wide_non_imputed[dmso.e_dmso.c_sig$symbol,],
                   fname = "DMSO.EIF2A_DMSO.Ctrl")
heatmap_proteomics(dat = protein_data_wide_non_imputed[tg.e_tg.c_sig$symbol,],
                   fname = "Tg.EIF2A_Tg.Ctrl")
heatmap_proteomics(dat = protein_data_wide_non_imputed[tg.c_dmso.c_sig$symbol,],
                   fname = "Tg.Ctrl_DMSO.Ctrl")
heatmap_proteomics(dat = protein_data_wide_non_imputed[tg.e_dmso.e_sig$symbol,],
                   fname = "Tg.EIF2A_DMSO.EIF2A")

# plot significant proteins, with imputed values
heatmap_proteomics(dat = protein_data_wide[dmso.e_dmso.c_sig$symbol,],
                   fname = "DMSO.EIF2A_DMSO.Ctrl", imputed = TRUE)
heatmap_proteomics(dat = protein_data_wide[tg.e_tg.c_sig$symbol,],
                   fname = "Tg.EIF2A_Tg.Ctrl", imputed = TRUE)
heatmap_proteomics(dat = protein_data_wide[tg.c_dmso.c_sig$symbol,],
                   fname = "Tg.Ctrl_DMSO.Ctrl", imputed = TRUE)
heatmap_proteomics(dat = protein_data_wide[tg.e_dmso.e_sig$symbol,],
                   fname = "Tg.EIF2A_DMSO.EIF2A", imputed = TRUE)

# plot reversed proteins
heatmap_proteomics(dat = protein_data_wide_non_imputed[dmso.c_tg.c_tg.e$symbol,],
                   fname = "DMSO.Ctrl_Tg.Ctrl_Tg.EIF2A")

heatmap_proteomics(dat = protein_data_wide[dmso.c_tg.c_tg.e$symbol,],
                   fname = "DMSO.Ctrl_Tg.Ctrl_Tg.EIF2A", imputed = TRUE)

heatmap_proteomics(dat = protein_data_wide_non_imputed[dmso.c_tg.c_tg.e_0.25$symbol,],
                   fname = "DMSO.Ctrl_Tg.Ctrl_Tg.EIF2A...0.25detected")

# plot top proteins with p<0.001
heatmap_proteomics(dat = protein_data_wide_non_imputed[dmso.e_dmso.c[dmso.e_dmso.c$P.Value<0.001,]$symbol,],
                   fname = "DMSO.EIF2A_DMSO.Ctrl p.001")
heatmap_proteomics(dat = protein_data_wide_non_imputed[tg.e_tg.c[dmso.e_dmso.c$P.Value<0.001,]$symbol,],
                   fname = "Tg.EIF2A_Tg.Ctrl p.001")

# heatmap function for EIF2A islet proteomics
heatmap_proteomics <- function(dat, fname, imputed=FALSE){
  # fname should specify the comparison separated by "_". E.g."Tg.EIF2A_Tg.Ctrl".
  m <- dat
  m.z <- t(scale(t(m))) # scale -- calculate Z-score for columns: [(x-mean)/SD]; t -- transpose the matrix; log2 -- log2 transform the data
  m.z[is.na(m.z)] <- 0
#rownames(m.z) <- rownames(m) # set gene names as rownames of m.z

# plot heatmap  --------------
# (originally established by Yi Han Xia)

number_dc <- length(grep("DMSO.ctrl", colnames(m.z))) # grep() -- find the position/index of, in this case, column names that contain the pattern "DN"
number_de <- length(grep("DMSO.EIF2A", colnames(m.z)))
number_tc <- length(grep("Tg.ctrl", colnames(m.z)))
number_te <- length(grep("Tg.EIF2A", colnames(m.z))) 

end_dc <- grep("DMSO.ctrl", colnames(m.z))[number_de] # get the end position/index of each group
end_de <- grep("DMSO.EIF2A", colnames(m.z))[number_de]
end_tc <- grep("Tg.ctrl", colnames(m.z))[number_tc]
end_te <- grep("Tg.EIF2A", colnames(m.z))[number_te]



#set.seed(12) # run this to get the same k-means clusters 


# heatmap for DMSO_EIF2A vs DMSO or Tg_EIF2A vs Tg                
heatmap <- Heatmap(m.z,
                   column_title = gsub("_", " vs ", fname),
                   
                   show_row_names = TRUE,#FALSE,
                   show_column_names = FALSE,
                   show_row_dend = TRUE,
                   show_column_dend = FALSE,
                   
                   clustering_distance_rows = "euclidean",
                   clustering_method_rows = "ward.D2", # default linkage method is "complete"
                   
                   row_dend_side = "left",
                   row_dend_width = unit(15, "mm"),
                   
                   column_names_side = "top",
                   column_dend_side = "bottom",
                   
                   row_labels = gt_render(rownames(m.z)), ### now we set row names (genes) as labels
                   row_names_gp = gpar(fontsize = 10),
                  
                   col = colorRamp2(c(min(na.omit(m.z)),0,max(na.omit(m.z))), c("blue", "white", "red")), # set the color range from Z-score -3 to 3 (saturate the outliers,  if you didn't remove them).
                   #col = colorRamp2(c(min(na.omit(m.z)),0,max(na.omit(m.z))), c("blue", "white", "red")), # alternatively, set the range from min to max value
                   
                   column_order = 1:ncol(m.z),
                   
                   height = nrow(m.z)*unit(3, "mm"), # set the height of heatmap
                   width = ncol(m.z)*unit(8, "mm"), # set the column width
                   
                   top_annotation = columnAnnotation(empty = anno_empty(border = FALSE, 
                                                                        height = unit(6, "mm"))), # leave an empty space on top for the condition labels below.
                   
                   border_gp = gpar(col = "black"), # set border color
                   
                   show_heatmap_legend = TRUE,
                   heatmap_legend_param = list(
                     title = "Row z-score",
                     title_position = "topleft",
                     #at = c(-3, -2, -1, 0, 1, 2, 3), # set ticks/numbers of legend
                     legend_height = unit(4, "cm")),
                   
                   #split = 4, # split the heatmap into separate blocks by hierarchical clustering
                   #row_km = 4, # or split the heatmap into separate blocks by k-means clustering
                   ## determine the number of clusters: https://www.datanovia.com/en/lessons/determining-the-optimal-number-of-clusters-3-must-know-methods/
                   
                  layer_fun = function(j, i, x, y, width, height, fill) {
                     mat = restore_matrix(j, i, x, y)
                     
                     ind = unique(c(mat[, c(end_dc,
                                            end_de,
                                            end_tc
                                            ) # enter the number where you want a gap between columns
                                        ])) 
                     
                     grid.rect(x = x[ind] + unit(0.5/ncol(m.z), "npc"), 
                               y = y[ind], 
                               width = unit(0.03, "inches"), # width of the gap
                               height = unit(1/nrow(m.z), "npc"),
                               gp = gpar(col = "white", fill = "white") # color of the gap
                     )
                   }
                   ) 

# save and label heatmap ---------------

#step 1 - save a pdf file
{
  if (imputed){
    pdf(file = paste0("figures/heatmap_EIF2A_proteomics_imputed_", fname, ".pdf"),
    width = 6.5, 
    height = nrow(m.z)*0.12 + 1)
  } else{
    pdf(file = paste0("figures/heatmap_EIF2A_proteomics_", fname, ".pdf"),
    width = 6.5, 
    height = nrow(m.z)*0.12 + 1)
  }
  

#step 2 - draw heatmap in the saved file
draw(heatmap)

#step 3 - draw the labeling on the heatmap in the saved file

# labeling heatmap
#list_components() # get the viewport names
seekViewport("annotation_empty_1") # seek the empty space we saved at the top of heatmap, called "annotation_empty_1"

## draw text labels
grid.text(expression("DMSO\nControl"), 
          x = (end_dc)/2/ncol(m.z), # location at the middle of DN
          y = 0.2,
          just = c("center", "center"),
          gp = gpar(fontsize = 11,
                    col = "black"))

grid.text(expression("DMSO\nEIF2A"), 
          x = (end_dc + end_de)/2/ncol(m.z),
          y = 0.2,
          just = c("center", "center"),
          gp = gpar(fontsize = 11))

grid.text(expression("Tg\nControl"), 
          x = (end_de + end_tc)/2/ncol(m.z), # location at the middle of DH
          y = 0.2,
          just = c("center", "center"),
          gp = gpar(fontsize = 11))

grid.text(expression("Tg\nEIF2A"), 
          x = (end_tc + end_te)/2/ncol(m.z),
          y = 0.2,
          just = c("center", "center"),
          gp = gpar(fontsize = 11))

dev.off()
}
}

```


```{r volcano proteomics}
dmso.e_dmso.c <- read_excel("output/EIF2A_proteomics_limma_DE_results_filtered.N2.xlsx", sheet = 1)
tg.e_tg.c <- read_excel("output/EIF2A_proteomics_limma_DE_results_filtered.N2.xlsx", sheet = 2)
tg.c_dmso.c <- read_excel("output/EIF2A_proteomics_limma_DE_results_filtered.N2.xlsx", sheet = 3)
tg.e_dmso.e <- read_excel("output/EIF2A_proteomics_limma_DE_results_filtered.N2.xlsx", sheet = 4)

colnames(dmso.e_dmso.c)

plot.volcano(df = dmso.e_dmso.c,
             p_col = 'adj.P.Val',
             p_cutoff = 0.05,
             log2FC_col = 'logFC',
             log2FC_cutoff = 0.5,
             gene_num = 30,
             gene_col = 'symbol',
             name = 'proteomics_DMSO.EIF2A_DMSO.Ctrl')

plot.volcano(df = tg.e_tg.c,
             p_col = 'adj.P.Val',
             p_cutoff = 0.05,
             log2FC_col = 'logFC',
             log2FC_cutoff = 0.5,
             gene_num = 30,
             gene_col = 'symbol',
             name = 'proteomics_Tg.EIF2A_Tg.Ctrl')

plot.volcano(df = tg.c_dmso.c,
             p_col = 'adj.P.Val',
             p_cutoff = 0.05,
             log2FC_col = 'logFC',
             log2FC_cutoff = 0.5,
             gene_num = 30,
             gene_col = 'symbol',
             name = 'proteomics_Tg.Ctrl_DMSO.Ctrl')


plot.volcano(df = tg.e_dmso.e,
             p_col = 'adj.P.Val',
             p_cutoff = 0.05,
             log2FC_col = 'logFC',
             log2FC_cutoff = 0.5,
             gene_num = 30,
             gene_col = 'symbol',
             name = 'proteomics_Tg.EIF2A_DMSO.EIF2A')

```


```{r volcano RNA-seq}
dmso.e_dmso.c <- read.csv("output/RNAseq_DESeq2_eIF2A0_control0_noOutlier_lfc0.5.csv")
tg.e_tg.c <- read.csv("output/RNAseq_DESeq2_eIF2A3_control3_noOutlier_lfc0.5.csv")
tg.c_dmso.c <- read.csv("output/RNAseq_DESeq2_control3_control0_noOutlier_lfc0.5.csv")
tg.e_dmso.e <- read.csv("output/RNAseq_DESeq2_eIF2A3_eIF2A0_noOutlier_lfc0.5.csv")
View(dmso.e_dmso.c)
colnames(dmso.e_dmso.c)

plot.volcano(df = dmso.e_dmso.c,
             p_col = 'padj',
             p_cutoff = 0.05,
             log2FC_col = 'log2FoldChange',
             log2FC_cutoff = 0.5,
             gene_num = 30,
             gene_col = 'symbol',
             name = 'RNAseq_EIF2A0_Ctrl0')

plot.volcano(df = tg.e_tg.c,
             p_col = 'padj',
             p_cutoff = 0.05,
             log2FC_col = 'log2FoldChange',
             log2FC_cutoff = 0.5,
             gene_num = 30,
             gene_col = 'symbol',
             name = 'RNAseq_EIF2A3_Ctrl3')

plot.volcano(df = tg.c_dmso.c,
             p_col = 'padj',
             p_cutoff = 0.05,
             log2FC_col = 'log2FoldChange',
             log2FC_cutoff = 0.5,
             gene_num = 30,
             gene_col = 'symbol',
             name = 'RNAseq_Ctrl3_Ctrl0')


plot.volcano(df = tg.e_dmso.e,
             p_col = 'padj',
             p_cutoff = 0.05,
             log2FC_col = 'log2FoldChange',
             log2FC_cutoff = 0.5,
             gene_num = 30,
             gene_col = 'symbol',
             name = 'RNAseq_EIF2A3_EIF2A0')

```
